tdrs-backend/tdpservice/scheduling/tasks.py

```diff
@@ -0,0 +1,65 @@
+from __future__ import absolute_import
+
+import os
+
+from celery import shared_task
+from django.conf import settings
+import datetime
+import paramiko
+import logging
+from tdpservice.data_files.models import DataFile
+logger = logging.getLogger(__name__)
+
+server_address = settings.SERVER_ADDRESS
+local_key = settings.LOCAL_KEY
+username = settings.USERNAME
+
+
+def write_key_to_file(private_key):
+    """
+    Paramiko require the key in file object format
+    """
+    with open('temp_key_file', 'w') as f:
+        f.write(private_key)
+        f.close()
+    return 'temp_key_file'
+
+
+@shared_task
+def upload(data_file_pk):
+    """
+    This task uploads the file in DataFile object with
+    pk = data_file_pk
+    to sftp server as defined in Settings file
+    """
+    data_file = DataFile.objects.get(id=data_file_pk)
+    destination = str(data_file.file).split('/')[-1]
+    today_date = datetime.datetime.today()
+    upper_directory_name = today_date.strftime('%Y%m%d')
+    lower_directory_name = today_date.strftime(str(data_file.year) + '-' + str(data_file.quarter))
+
+    transport = paramiko.SSHClient()
+    transport.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+    transport.connect(server_address, key_filename=write_key_to_file(local_key), username=username)
+    sftp = transport.open_sftp()
+
+    transport.exec_command('mkdir -p ' + upper_directory_name +
+                           '/' + lower_directory_name)
+
+    f = data_file.file.read()
+    with open(destination, 'wb') as f1:
+        f1.write(f)
+        f1.close()
+    logger.info(os.listdir())
+
+    # Upload file
+    try:
+        # sftp = transport.open_sftp()
+        # temp file can be the file object
+        sftp.put(destination, upper_directory_name + '/' + lower_directory_name + '/' + destination)
+        os.remove(destination)
```

Comment:
If a huge amount of small files goes through here something like mmap might be worth it. mmap maps files to memory, but allows them to still behave like file objects in most cases. I am not sure if it would work with sftp, but it might be an easy drop in.

Also, is this an asynchronous task? If two users did this at the same time with the same set of files would it be ok? 

https://docs.python.org/3/library/mmap.html

Reply:
File sizes can vary. 

Since this is asynchronous the celery task scheduler is able to handle more than 500k tasks schedules/sec, however, the tasks stay in queue until there is enough processing power available.